src/data/list/basic.lean

```diff
@@ -4552,6 +4552,33 @@ theorem prod_range_succ {α : Type u} [monoid α] (f : ℕ → α) (n : ℕ) :
 by rw [range_concat, map_append, map_singleton,
   prod_append, prod_cons, prod_nil, mul_one]
 
+@[to_additive]
+theorem prod_range_succ' {α : Type u} [monoid α] (f : ℕ → α) (n : ℕ) :
+  ((range n.succ).map f).prod = f 0 * ((range n).map (λ i, f (succ i))).prod :=
+begin
+  induction n with d hd,
+  { show 1 * f 0 = f 0 * 1, rw [one_mul, mul_one]},
+  { rw [list.prod_range_succ, hd, mul_assoc, ←list.prod_range_succ]},
+end
+
+theorem sum_map_mul_left {α : Type u} [semiring α] {β : Type*} (L : list β)
+  (f : β → α) (r : α) :
+  (L.map (λ b, r * f b)).sum = r * (L.map f).sum :=
+begin
+  induction L with b L hL,
+  { exact (mul_zero r).symm},
+  { simp [mul_add, hL]}
+end
+
+theorem sum_map_mul_right {α : Type*} [semiring α] {β : Type*} (L : list β)
+  (f : β → α) (r : α) :
+  (L.map (λ b, f b * r)).sum = (L.map f).sum * r :=
+begin
+  induction L with b L hL,
+  { exact (zero_mul r).symm},
+  { simp [add_mul, hL]}
+end
```

Comment:
Could we factor out a proof of 
```
@[to_additive]
theorem prod_map_hom {α β γ : Type*} [monoid β] [monoid γ] (L : list α) (f : α → β) (g : β →* γ) :
  (L.map (g ∘ f)).prod = g ((L.map f).prod) := sorry
```
and then use the fact that scalar multiplication in a semiring is an `→+`? Even without refactoring the proofs of these two statements it would be nice to have the general result.

Reply:
(Feel free to say: "go ahead". :-)

Reply:
Instead I'll say "why doesn't this work?"
```
theorem prod_map_hom {α β γ : Type*} [monoid β] [monoid γ] (L : list α) (f : α → β) (g : β →* γ) :
  (L.map (g ∘ f)).prod = g ((L.map f).prod) :=
L.rec g.map_one.symm (λ _ _ IH, by simp [g.map_mul, IH])
```

This works:
```
@list.rec _ (λ L, (L.map (g ∘ f)).prod = g ((L.map f).prod)) g.map_one.symm (λ _ _ IH, by simp [g.map_mul, IH]) L
```

I've used a tactic mode proof.

Reply:
Re: the application. Do you think the construction of an add_monoid_hom R -> R from left multiplication by an element of R should be added earlier than [here](https://github.com/leanprover-community/mathlib/blob/003141c88763fd628c8a728402f79b5bddf3a639/src/algebra/module.lean#L266-L267)? Note the "move" comment directly afterwards.

OK, talking about it [here](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/multiplication.20is.20an.20add_monoid_hom) I'm just going to add the constructions myself.